File: 1299-replace-elements-with-greatest-element-on-right-side/1299-replace-elements-with-greatest-element-on-right-side.cpp
//time:o(n)
//space:o(1)

class Solution {
public:
    vector<int> replaceElements(vector<int>& arr) {
        
        int n = arr.size();
        int max = arr[n-1];
        arr[n-1] = -1;
        
        for(int i=n-2;i>=0;i--){
            int temp = max;
            if(max < arr[i]){
                max = arr[i];
            }
            arr[i] = temp;
        }
        return arr;
        
            
    }
};

/*
int main()
{
Solution solution;
vector<int> arr = {8,5,6,7,5};
vector<int> result = solution.replaceElements(arr);

for(int i:result){
cout<<i<<endl;
}
cout<<endl;
return 0;
}
*/

File: 1299-replace-elements-with-greatest-element-on-right-side/1299-replace-elements-with-greatest-element-on-right-side.py
class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        
        #initial max = -1
        #reverse iteration
        #new max = max(old max , arr[i])
        
        rightmax = -1
        
        for i in range(len(arr)-1 , -1 ,-1):
            newmax = max(rightmax,arr[i])
            arr[i] = rightmax
            rightmax = newmax
            
        return arr    
    
    
#solution = Solution()

#arr = [8,5,4,5,1]

#result = solution.replaceElements(arr)

#print(result)

        
        
        